src/render/ngage/SDL_render_ngage.cpp

```diff
@@ -0,0 +1,790 @@
+{
+    if (iRenderer && iRenderer->Gc())
+    {
+        TRect viewportRect(aX, aY, aX + aWidth, aY + aHeight);
+        iRenderer->Gc()->SetClippingRect(viewportRect);
+    }
+}
+
+void CRenderer::UpdateFPS()
+{
+    static TTime lastTime;
+    static TInt frameCount = 0;
+    TTime currentTime;
+    const TUint KOneSecond = 1000000; // 1s in ms.
+
+    currentTime.HomeTime();
+    ++frameCount;
+
+    TTimeIntervalMicroSeconds timeDiff = currentTime.MicroSecondsFrom(lastTime);
+
+    if (timeDiff.Int64() >= KOneSecond)
+    {
+        // Calculate FPS.
+        iFPS = frameCount;
+
+        // Reset frame count and last time.
+        frameCount = 0;
+        lastTime = currentTime;
+    }
+}
+
+static SDL_Scancode ConvertScancode(int key)
+{
+    SDL_Keycode keycode;
+
+    switch(key)
+    {
+        case EStdKeyBackspace: // Clear key
+            keycode = SDLK_BACKSPACE;
+            break;
+        case 0x31: // 1
+            keycode = SDLK_1;
+            break;
+        case 0x32: // 2
+            keycode = SDLK_2;
+            break;
+        case 0x33: // 3
+            keycode = SDLK_3;
+            break;
+        case 0x34: // 4
+            keycode = SDLK_4;
+            break;
+        case 0x35: // 5
+            keycode = SDLK_5;
+            break;
+        case 0x36: // 6
+            keycode = SDLK_6;
+            break;
+        case 0x37: // 7
+            keycode = SDLK_7;
+            break;
+        case 0x38: // 8
+            keycode = SDLK_8;
+            break;
+        case 0x39: // 9
+            keycode = SDLK_9;
+            break;
+        case 0x30: // 0
+            keycode = SDLK_0;
+            break;
+        case 0x2a: // Asterisk
+            keycode = SDLK_ASTERISK;
+            break;
+        case EStdKeyHash: // Hash
+            keycode = SDLK_HASH;
+            break;
+        case EStdKeyDevice0: // Left softkey
+            keycode = SDLK_SOFTLEFT;
+            break;
+        case EStdKeyDevice1: // Right softkey
+            keycode = SDLK_SOFTRIGHT;
+            break;
+        case EStdKeyApplication0: // Call softkey
+            keycode = SDLK_CALL;
+            break;
+        case EStdKeyApplication1: // End call softkey
+            keycode = SDLK_ENDCALL;
+            break;
+        case EStdKeyDevice3: // Middle softkey
+            keycode = SDLK_SELECT;
+            break;
+        case EStdKeyUpArrow: // Up arrow
+            keycode = SDLK_UP;
+            break;
+        case EStdKeyDownArrow: // Down arrow
+            keycode = SDLK_DOWN;
+            break;
+        case EStdKeyLeftArrow: // Left arrow
+            keycode = SDLK_LEFT;
+            break;
+        case EStdKeyRightArrow: // Right arrow
+            keycode = SDLK_RIGHT;
+            break;
+        default:
+            keycode = SDLK_UNKNOWN;
+            break;
+    }
+
+    return SDL_GetScancodeFromKey(keycode, NULL);
+}
+
+void CRenderer::HandleEvent(const TWsEvent& aWsEvent)
```

Comment:
Would it be possible to separate event handling from the rendering code?

Reply:
Speaking of code placement: In my example application I also pointed out that SDLs audio sub-system should be initialised, even when it is not needed. This ensures that all resources are properly released when the application is closed.

Why? Because the rendering and audio backends are initialised early in `E32Main()` before the active scheduler is activated. There may be other ways to do this, but it's not easy to find good example code for such an old platform.